### ISSUE TITLE
Maybe withTempDirectory should ignore permission errors:

### ISSUE THREAD

**Issue by ezyang** [haskell/cabal#3308]
The following test program, compiled against `temporary-1.2.0.4` and `Cabal-1.24.0.0` (specifically a branch derived off of 2f87646e2867a916b74a9fdd30adba45837b8e36) behaves differently:

```
import System.FilePath
import Control.Concurrent
import System.Directory
import Distribution.Verbosity

import qualified System.IO.Temp as A
import qualified Distribution.Simple.Utils as B

main = do
    tmp <- getTemporaryDirectory
    (r,fn) <- A.withTempDirectory "tmp" tmp $ \path -> do
        let fn = path </> "foo"
        writeFile fn "1\n2\n3\n"
        r <- fmap lines (readFile fn)
        return (r, fn)
    print =<< doesFileExist fn
    print r

    (r,fn) <- B.withTempDirectory normal tmp "tmp" $ \path -> do
        let fn = path </> "foo"
        writeFile fn "1\n2\n3\n"
        r <- fmap lines (readFile fn)
        return (r, fn)
    print =<< doesFileExist fn
    print r
```

It's unclear if you'd expect this code to pass or fail (it always passes on Linux, so perhaps that's better). But actually, the code using `temporary` succeeds and our code fails!

```
Edward@glass MSYS ~
$ runghc TmpTest.hs
True
["1","2","3"]
TmpTest.hs: DeleteFile "C:\\msys64\\tmp\\tmp-11944\\foo": permission denied (The process cannot access the file because it is being used by another process.)
```

The reason is that `temporary` suppresses all IO errors:

```

-- | Create and use a temporary directory.
--
-- Creates a new temporary directory inside the given directory, making use
-- of the template. The temp directory is deleted after use. For example:
--
-- > withTempDirectory "src" "sdist." $ \tmpDir -> do ...
--
-- The @tmpDir@ will be a new subdirectory of the given directory, e.g.
-- @src/sdist.342@.
withTempDirectory :: (MonadMask m, MonadIO m) =>
                     FilePath -- ^ Temp directory to create the directory in
                  -> String   -- ^ Directory name template. See 'openTempFile'.
                  -> (FilePath -> m a) -- ^ Callback that can use the directory
                  -> m a
withTempDirectory targetDir template =
  Exception.bracket
    (liftIO (createTempDirectory targetDir template))
    (liftIO . ignoringIOErrors . removeDirectoryRecursive)
```

whereas we only ignore file does not exist errors. Maybe we should ignore all errors too?

Not ignoring permission errors resulted in #3306.

**Comment by 23Skidoo** [comment]
I kind of like that it fails loudly, forcing us to fix the error properly. OTOH, your suggestion does make for better user experience, which is probably more important.

**Comment by ezyang** [comment]
Two problems:

1. The error message can easily become something obtuse. We got something good here, but when I was debugging #3306 all I got for an error was "cabal.exe: permission denied". That's really bad! (On reflection, I am not sure why this was the case. Maybe it's the fault of `annotateFailure`.
2. Arguably, keeping the temporary file around is the desirable state of affairs:then we can lazily read in the contents rather than eagerly strictly read it in. In Linux we get the best of both worlds: the file is "removed" but we'll keep a handle to it until we're done reading it in (and ostensibly are finished using its results.)

**Comment by ezyang** [comment]
Actually, (1) is due to #3310. I'm going to submit a fix.

**Comment by 23Skidoo** [comment]
I'd prefer to minimise the use of lazy I/O, not expand it. Arguably the only place where the use of lazy I/O is motivated is reading the package index.

**Comment by ezyang** [comment]
OK, I'll fix #3310 and WONTFIX this.
